### PR TITLE
Add QgsAbstractDatabaseProviderConnection interface for renaming fields

### PR DESCRIPTION
**Pull request by nyalldawson**
And implement for OGR connections. Also expose the ability to rename existing fields for supported database connections via browser.

### CODE REVIEW

File: src/core/providers/ogr/qgsogrproviderconnection.cpp [hunk 1]
```diff
@@ -756,6 +757,53 @@ void QgsOgrProviderConnection::addFieldDomain( const QgsFieldDomain &domain, con
 #endif
 }
 
+void QgsOgrProviderConnection::renameField( const QString &schema, const QString &tableName, const QString &name, const QString &newName ) const
+{
+  checkCapability( Capability::RenameField );
```
Comment (nirvn) [inline src/core/providers/ogr/qgsogrproviderconnection.cpp:762]: What is this line actually supposed to do here? shouldn't it be something like if ( !checkCapability() ) return; ?

File: src/core/providers/qgsabstractdatabaseproviderconnection.cpp [hunk 2]
```diff
@@ -1184,6 +1184,38 @@ void QgsAbstractDatabaseProviderConnection::addField( const QgsField &field, con
   }
 }
 
+void QgsAbstractDatabaseProviderConnection::renameField( const QString &schema, const QString &tableName, const QString &name, const QString &newName ) const
+{
+  checkCapability( Capability::RenameField );
```
Comment (nirvn) [inline src/core/providers/qgsabstractdatabaseproviderconnection.cpp:1189]: Same question as above.

File: tests/src/python/test_qgsproviderconnection_ogr_gpkg.py [hunk 3]
```diff
@@ -322,6 +322,34 @@ def test_execute_sql_pk_geoms(self):
...
+            conn.renameField('schema', 'cdb_lines', 'asd', 'xyz')
+        # try to rename over existing field
+        with self.assertRaises(QgsProviderConnectionException):
+            conn.renameField('schema', 'cdb_lines', 'name', 'ortsrat')
+        # try to rename geometry field
+        with self.assertRaises(QgsProviderConnectionException):
+            conn.renameField('schema', 'cdb_lines', 'geom', 'the_geom')
+
+        # good rename
+        conn.renameField('schema', 'cdb_lines', 'name', 'name2')
```
Comment (nirvn) [inline tests/src/python/test_qgsproviderconnection_ogr_gpkg.py:349]: For geopackages, do we _need_ to provide a schema string or could it be blank? Either way probably worth testing out.

File: src/core/providers/ogr/qgsogrproviderconnection.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/core/providers/ogr/qgsogrproviderconnection.cpp:762]: It throws an exception if the capability isn't supported, so there's no need to handle anything manually

File: tests/src/python/test_qgsproviderconnection_ogr_gpkg.py (hunk 3 above)
Comment (nyalldawson) [reply tests/src/python/test_qgsproviderconnection_ogr_gpkg.py:349]: It's completely ignored

Comment (nirvn) [reply tests/src/python/test_qgsproviderconnection_ogr_gpkg.py:349]: @nyalldawson , maybe add an empty string test then? It's also useful for people who would look at test code to try and familiarize themselves with the API.

File: src/core/providers/ogr/qgsogrproviderconnection.cpp (hunk 1 above)
Comment (nirvn) [reply src/core/providers/ogr/qgsogrproviderconnection.cpp:762]: @nyalldawson , ah, I was wondering, good good.

File: tests/src/python/test_qgsproviderconnection_ogr_gpkg.py (hunk 3 above)
Comment (nyalldawson) [reply tests/src/python/test_qgsproviderconnection_ogr_gpkg.py:349]: Done

**Review by nirvn — APPROVED** [review]

**Pull request merged**